source/lac/petsc_parallel_sparse_matrix.cc

```diff
@@ -890,6 +890,48 @@ namespace PETScWrappers
       return u*tmp;
     }
 
+    IndexSet
+    SparseMatrix::locally_owned_domain_indices () const
+    {
+      PetscInt n_rows, n_cols, min, max;
+      PetscErrorCode ierr;
+      IS *rows = nullptr;
+      IS *cols = nullptr;
+
+      ierr = MatGetSize (matrix, &n_rows, &n_cols);
+      AssertThrow (ierr == 0, ExcPETScError(ierr));
+      ierr = MatGetOwnershipIS(matrix, rows, cols);
+      AssertThrow (ierr == 0, ExcPETScError(ierr));
+      ierr = ISGetMinMax(*rows, &min, &max);
+      AssertThrow (ierr == 0, ExcPETScError(ierr));
+
+      IndexSet locally_owned_domain_indices(n_rows);
+      locally_owned_domain_indices.add_range(min, max);
```

Comment:
Good question. You could presumably ask `ISGetLocalSize` on the `rows` object here and make sure it equals `max-min+1`. That may make sense.

I looked at the PETSc index set pages at http://www.mcs.anl.gov/petsc/petsc-current/docs/manualpages/IS/index.html and realize that it's not the index set itself that is limited to contiguous ranges, but the matrices. The index sets are quite general, and you could ask it whether it's contiguous. But my understanding is that the index set you get from a matrix will always be (at least for now; the future is hard to predict).